jcommune-service/src/main/java/org/jtalks/jcommune/service/security/AclGroupPermissionEvaluator.java

```diff
@@ -340,10 +341,14 @@ private boolean isRestrictedForGroup(List<GroupAce> controlEntries,
     private boolean isGrantedForGroup(List<GroupAce> controlEntries,
                                       Authentication authentication, Permission permission,
                                       boolean isCheckAllowedGrant) {
-        if (authentication.getPrincipal() instanceof JCUser) {
+        Object auth = authentication.getPrincipal();
+        if (auth instanceof JCUser) {
+            List<Long> groupsIDs = ((JCUser) auth).getGroupsIDs();
```

Comment:
Hypothetically (!) can be improved with a HashSet since its complexity is O(1) instead of O(n). But that depends on the size of `groupIDs` and `controlEntries` as if they are small then building a HashSet is probably going to take more time than to iterate over the arrays multiple times.